airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/poetry/publish/pipeline.py

```diff
@@ -0,0 +1,153 @@
+from pipelines.consts import PYPROJECT_TOML_FILE_PATH, SETUP_PY_FILE_PATH
+from pipelines.dagger.actions.python.poetry import with_poetry
+from pipelines.helpers.utils import sh_dash_c
+from pipelines.models.steps import Step, StepResult
+
+
+class PackageType(Enum):
+    POETRY = auto()
+    PIP = auto()
+    NONE = auto()
+
+
+class PublishToPyPI(Step):
+    context: PyPIPublishContext
+    title = "Publish package to PyPI"
+
+    def _get_base_container(self) -> Container:
+        return with_poetry(self.context)
+
+    async def _get_package_metadata_from_pyproject_toml(self, dir_to_publish: Directory) -> Optional[Tuple[str, str]]:
+        pyproject_toml = dir_to_publish.file(PYPROJECT_TOML_FILE_PATH)
+        pyproject_toml_content = await pyproject_toml.contents()
+        contents = tomli.loads(pyproject_toml_content)
+        if (
+            "tool" not in contents
+            or "poetry" not in contents["tool"]
+            or "name" not in contents["tool"]["poetry"]
+            or "version" not in contents["tool"]["poetry"]
+        ):
+            return None
+
+        return (contents["tool"]["poetry"]["name"], contents["tool"]["poetry"]["version"])
+
+    async def _get_package_type(self, dir_to_publish: Directory) -> PackageType:
+        files = await dir_to_publish.entries()
+        has_pyproject_toml = PYPROJECT_TOML_FILE_PATH in files
+        has_setup_py = SETUP_PY_FILE_PATH in files
+        if has_pyproject_toml:
+            return PackageType.POETRY
+        elif has_setup_py:
+            return PackageType.PIP
+        else:
+            return PackageType.NONE
+
+    async def _run(self) -> StepResult:
+        dir_to_publish = await self.context.get_repo_dir(self.context.package_path)
+        package_type = await self._get_package_type(dir_to_publish)
+
+        if package_type == PackageType.NONE:
+            return self.skip("Connector does not have a pyproject.toml file or setup.py file, skipping.")
+
+        # Try to infer package name and version from the pyproject.toml file. If it is not present, we need to have the package name and version set
+        # Setup.py packages need to set package name and version as parameter
+        if not self.context.package_name or not self.context.version:
+            if not package_type == PackageType.POETRY:
+                return self.skip(
+                    "Connector does not have a pyproject.toml file and version and package name is not set otherwise, skipping."
+                )
+
+            package_metadata = await self._get_package_metadata_from_pyproject_toml(dir_to_publish)
+
+            if not package_metadata:
+                return self.skip(
+                    "Connector does not have a pyproject.toml file which specifies package name and version and they are not set otherwise, skipping."
+                )
+
+            self.context.package_name = package_metadata[0]
+            self.context.version = package_metadata[1]
```

Comment:
```suggestion
            self.context.package_metadata : PackageMetadata = package_metadata
```
nit: Just suggesting bundling package metadata inside a new dataclass.